### PR TITLE
fix concat

### PR DESCRIPTION
**Pull request by dorlib**
not the prittiest buts seems to be working, 
I would be glad for improvement suggestions

### CODE REVIEW

File: avl_template_new.py [hunk 1]
```diff
@@ -1072,21 +1072,45 @@ def shuffle(self, array):
...
 	"""
 	def concat(self, lst):
-		absDiff = abs(self.root.height - lst.root.height)
+		absDiff = abs(self.getRoot().getHeight() - lst.getRoot().getHeight())
+		shorterTree = self
+		LongerTree = lst
+
+		b = LongerTree.getRoot()
+		while b.getHeight() > shorterTree.getRoot().getHeight():
+			b = b.getLeft()
```
Comment (ghost) [inline avl_template_new.py:1084]: what if b has only a right child?

File: avl_template_new.py [hunk 2]
```diff
@@ -1072,21 +1072,45 @@ def shuffle(self, array):
...
 		
-		LastNodeInSelf = self.retrieveNode(self.size)
-		firstNodeInLst = lst.retrieveNode(self.size)
+		maxInShort = shorterTree.retrieveNode(shorterTree.size - 1)
+		maxInShort.getParent().right = AVLNode.virtualNode()
```
Comment (ghost) [inline avl_template_new.py]: 1. you can use `shortTree.getMaximum()` instead
2. you can use `shortTree.deleteLeaf(maxInShort)` which would be more readable

File: avl_template_new.py [hunk 3]
```diff
@@ -1072,21 +1072,45 @@ def shuffle(self, array):
...
+
+		# if be is the root of lst
```
Comment (ghost) [inline avl_template_new.py]: comments are flipped, you write "if b is the root" but the condition is true iff b is not the root.

File: avl_template_new.py [hunk 4]
```diff
@@ -1072,21 +1072,45 @@ def shuffle(self, array):
...
+		if b.getParent() != None:
+			maxInShort.parent = b.getParent()
+			maxInShort.getParent().left = maxInShort
+
+			maxInShort.right = b
+			b.parent = maxInShort
+
+			maxInShort.left = shorterTree.getRoot()
+			maxInShort.setHeight(max(maxInShort.getRight().getHeight(), maxInShort.getLeft().getHeight())+1)
```
Comment (ghost) [inline avl_template_new.py]: you can use `maxInShort.updateHelpers()` and it will also set the size correctly

File: avl_template_new.py [hunk 5]
```diff
@@ -1072,21 +1072,45 @@ def shuffle(self, array):
...
+
+		# if b is not the root of lst
+		else:
+			maxInShort.parent = None
+			maxInShort.right = b
+			b.parent = maxInShort
 
-		LastNodeInSelf.right = firstNodeInLst
+			maxInShort.left = shorterTree.getRoot()
+			maxInShort.setHeight(max(maxInShort.getRight().getHeight(), maxInShort.getLeft().getHeight())+1)
```
Comment (ghost) [inline avl_template_new.py]: use `updateHelpers()`

**Review by ghost — COMMENTED** [review]
This looks great, I added some comments where I think you could call other functions and simplify the code and maybe handle some edge cases.

I'm not sure if this works for when `lst` has a bigger height then `self`, did you check it? 
The variable names appear to assume that `self` is shorter in height then `lst`.

File: avl_template_new.py (hunk 1 above)
Comment (dorlib) [reply avl_template_new.py:1084]: we are going over the laft side of the tree only, the algo will know how to deal if the original list is is the bigger tree

Comment (ghost) [reply avl_template_new.py:1084]: Okay, great

**Comment by dorlib** [comment]
at the begining I checked if the heights are different and maybe switch the variables, but then I realized its redundent because the if else after that will handle it properly anyway, I tested it when lst.length is greater and lesser that self.length and it works, I agree it confusing but we can say that we assume that and seport both, i can also give other names, i think thats what i will do.

**Pull request merged**